timely/src/dataflow/operators/core/input.rs

```diff
@@ -274,26 +357,48 @@ impl<T: Timestamp, C: Container> Handle<T, C> {
         self.pushers.push(pusher);
     }
 
-    // flushes our buffer at each of the destinations. there can be more than one; clone if needed.
-    #[inline(never)]
+    /// Extract all ready contents from the builder and distribute to downstream operators.
+    #[inline]
+    fn extract(&mut self) {
+        while let Some(container) = self.builder.extract() {
+            Self::flush_container(container, &mut self.buffer, &mut self.pushers, &self.now_at);
+        }
+    }
+
+    /// Flush all contents and distribute to downstream operators.
+    #[inline]
     fn flush(&mut self) {
-        for index in 0 .. self.pushers.len() {
-            if index < self.pushers.len() - 1 {
-                self.buffer2.clone_from(&self.buffer1);
-                Message::push_at(&mut self.buffer2, self.now_at.clone(), &mut self.pushers[index]);
-                debug_assert!(self.buffer2.is_empty());
+        while let Some(container) = self.builder.finish() {
+            Self::flush_container(container, &mut self.buffer, &mut self.pushers, &self.now_at);
+        }
+    }
+
+    /// flushes our buffer at each of the destinations. there can be more than one; clone if needed.
+    /// Does not take `self` because `flush` and `extract` borrow `self` mutably.
+    #[inline(never)]
+    fn flush_container(
+        container: &mut CB::Container,
+        buffer: &mut CB::Container,
+        pushers: &mut [Counter<T, CB::Container, Tee<T, CB::Container>>],
+        now_at: &T
+    ) {
+        for index in 0 .. pushers.len() {
+            if index < pushers.len() - 1 {
+                buffer.clone_from(container);
+                Message::push_at(buffer, now_at.clone(), &mut pushers[index]);
+                debug_assert!(buffer.is_empty());
```

Comment:
Seems maybe stronger than strictly needed, as `buffer` could be holding resources that might be useful. Perhaps a container should be able to hold resources even across a `clear()` call, and the assert can pass, but e.g. a `Vec<String>` might not want to be cleared, so that `clone_from(container)` can use the string allocations themselves. Or not, maybe too complicated.